---
title: "<span style='font-size: 28px'>Single-cell RNAseq of mouse heart development</style>"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(scran)
library(dynamicTreeCut)
library(RColorBrewer)
library(ggplot2)
library(ggpubr)

dir <- "/Users/ibarra01/OneDrive - CRUK Cambridge Institute/github/mouseHeart_earlyDev_atlas/"

palette(brewer.pal(n=12, "Set3"))
```

### Clustering

We start from the batch corrected, normalised counts produced in the `02_batchCorrection.Rmd` script.

```{r loadData}
## normalised, batch corrected counts
sce.corr <- readRDS(paste0(dir, "data/sce.goodQual.NORM.batchCorrected.Rds"))

## HVGs
hvgs <- read.table(paste0(dir, "results/HVGs_minMean1_FDR0.05.tsv"), stringsAsFactors = FALSE)
hvgs <- hvgs$V1

## UMAP
umap <- read.table(paste0(dir, "results/umapCoords_corrected.tab"))
```

The next step is to cluster the cells to define the cellular diversity of the dataset. We use the distances between cells based on expression of HVGs on the batch corrected data. Clusters are defined through hierarchical clustering and a dynamic tree cut. 

```{r cluster}
## use the distance between cells to identify clusters
dat <- assay(sce.corr)[hvgs,]
test.dist <- dist(t(dat))

## define clusters by hierarchical clustering and dynamic tree cut
test.clust <- hclust(test.dist, method="average")
cut <- cutreeDynamic(test.clust, distM=as.matrix(test.dist), minClusterSize=40, method="hybrid", deepSplit = 1, verbose = 0)
sce.corr$cluster <- cut

names(cut) <- colnames(dat)
write.table(cut, paste0(dir, "results/clusters_average_min40.tsv"), quote = FALSE, sep="\t", col.names = FALSE)

stopifnot(identical(names(cut), row.names(umap)))
umap$cluster <- cut
o <- order(umap$cluster)
plot(umap$x[o], umap$y[o], pch=16, cex=0.75, col=umap$cluster[o], xlab="UMAP - dim1", ylab="UMAP - dim2", bty="l")
```

This procedure results in `r length(unique(cut[cut!=0]))` clusters.

```{r outliers}
table(cut)[-1]
```

Importantly, clusters are comprised of cells from all different batches. 

```{r batchContrib}
tmp <- sce.corr[,-which(sce.corr$cluster==0)]
table(batch=tmp$batch, cluster=tmp$cluster)
# barplot(t(t(table(batch=tmp$batch, cluster=tmp$cluster))/colSums(table(batch=tmp$batch, cluster=tmp$cluster))*100), col=1:12)
```

Finally, we check the QC statistics to make sure that there isn't a cluster that behaves abnormally.

```{r qc, fig.width=10, fig.height=10}
qc <- read.table(paste0(dir, "data/QCstats_allCells.tsv"))
qc <- qc[names(cut[cut!=0]),]

par(mfrow=c(2,2))
boxplot(log10(qc$libSize)~cut[cut!=0], las=2, xlab="cluster", ylab=expression('log'[10]*' library size'), col=1:12)
boxplot(qc$nGenes/1e3~cut[cut!=0], las=2, xlab="cluster", ylab="number of genes expressed x 1000", col=1:12)
boxplot(qc$mit/qc$libSize*100~cut[cut!=0], las=2, xlab="cluster", ylab="% in MT", col=1:12)
boxplot(qc$ercc/qc$libSize*100~cut[cut!=0], las=2, xlab="cluster", ylab="% in ERCC", col=1:12)
```

### Cluster markers {.tabset}

To get an initial idea of the identity of each cluster, we use the `findMarkers` function to identify genes with large fold-changes in a particular cluster compared to the rest. We require genes to be significant in all comparisons against all other clusters; this returns genes that are most specific to each cluster. But can be problematic if there are some closely related clusters that share markers.

This approach returns significant genes for all but one cluster.

```{r DEanalysis}
## use normalised counts and block by batch instead of using batch corrected counts
sce <- readRDS(paste0(dir, "data/sce_goodQual.NORM.Rds"))

## add cluster info
stopifnot(identical(colnames(logcounts(sce)), names(cut)))
sce$clusters <- cut

## remove outlier cell
sce <- sce[,sce$clusters>0]

## find markers
keep <- rowMeans(logcounts(sce)) > 0.1
markersDE <- findMarkers(sce, groups = sce$clusters, block=sce$batch, direction="up", subset.row=keep, pval.type="all")
unlist(lapply(markersDE, function(x) sum(x$FDR<0.05)))

saveRDS(markersDE, file=paste0(dir, "results/markerGenes_pval_all.Rds"))
```

To recover markers for cluster 2, we instead require the gene to be significantly different against at least 8 of the 12 clusters. This of course returns higher numbers of significant genes.

```{r DEanalysis_soft}
markersDE.some <- findMarkers(sce, groups = sce$clusters, block=sce$batch, direction="up", subset.row=keep, pval.type="some", min.prop=0.75)
unlist(lapply(markersDE.some, function(x) sum(x$FDR<0.05)))

saveRDS(markersDE, file=paste0(dir, "results/markerGenes_pval_some0.75.Rds"))
```

Below is the expression of the top 10 genes found for each cluster:

```{r plot}
th <- theme_bw() + theme(axis.text.x = element_text(size=12), axis.title.x = element_text(size=12), axis.text.y = element_text(size=12), axis.title.y = element_text(size=12), axis.ticks.x = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"), panel.border = element_blank(), plot.title = element_text(face="bold", hjust = 0.5))

plotGeneOnUMAP <- function(umap=umap, data=sce, clusters=clusters, gene=gene){
  df <- data.frame(x=umap$x, y=umap$y, expr=as.numeric(logcounts(sce[gene,])), cluster=clusters)
  df <- df[df$cluster>0,]
  df <- df[order(df$expr),]
  
  p <- list()
  p[[1]] <- ggplot(df, aes(x,y)) + geom_point(aes(colour=expr), alpha = 0.5) + scale_colour_gradientn(colours = colorRampPalette(c("grey", "lightblue", "dodgerblue4", "royalblue4"))(100)) + ggtitle(rowData(sce)[id,1]) + xlab("") + ylab("") + labs(colour=expression('log'[2]*' counts')) + th + theme(legend.position = "none") + guides(colour = guide_colorbar(title.position = "bottom"))
  p[[2]] <- ggplot(df, aes(as.factor(cluster), expr)) + geom_boxplot(fill=1:12) + ggtitle(rowData(sce)[id,1]) + xlab("") + ylab("") + th
  return(p)
}
```

#### Cluster 1

```{r cluster1, fig.width=6, fig.height=3}
cluster <- 1
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 2

```{r cluster2, fig.width=6, fig.height=3}
cluster <- 2
for(i in 1:10){
  id <- row.names(markersDE.some[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 3

```{r cluster3, fig.width=6, fig.height=3}
cluster <- 3
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 4

```{r cluster4, fig.width=6, fig.height=3}
cluster <- 4
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 5

```{r cluster5, fig.width=6, fig.height=3}
cluster <- 5
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 6

```{r cluster6, fig.width=6, fig.height=3}
cluster <- 6
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 7

```{r cluster7, fig.width=6, fig.height=3}
cluster <- 7
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 8

```{r cluster8, fig.width=6, fig.height=3}
cluster <- 8
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 9

```{r cluster9, fig.width=6, fig.height=3}
cluster <- 9
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 10

```{r cluster10, fig.width=6, fig.height=3}
cluster <- 10
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 11

```{r cluster11, fig.width=6, fig.height=3}
cluster <- 11
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

#### Cluster 12

```{r cluster12, fig.width=6, fig.height=3}
cluster <- 12
for(i in 1:10){
  id <- row.names(markersDE[[cluster]])[i]
  plots <- plotGeneOnUMAP(umap = umap[-which(cut==0),], data = sce, clusters = cut[cut>0], gene = id)
  print(ggarrange(plotlist = plots, ncol=2, nrow=1))
}
```

###

```{r info}
sessionInfo()
```
