---
title: "<span style='font-size: 28px'>Single-cell RNAseq of mouse heart development</style>"
date: '23 May, 2020'
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
---



### Single-cell data of reference mesoderm regions

We have manually micro-dissected and sequenced the transcriptomes of cells from four anatomically defined reference regions of mesoderm: ventral cardiac mesoderm (CaM), cranial mesoderm (CrM), dorsal mesoderm (DM) and presomitic mesoderm (PSM). These `reference` cells with known anatomical location can be used to better understand the heterogeneity observed in the cardiac mesoderm progenitor populations captured in the `unbiased` dataset. 

---

In the paper, we refined the annotation of these four regions to better reflect the anatomical locations dissected. Annotations in [DataS4](https://science.sciencemag.org/highwire/filestream/755147/field_highwire_adjunct_files/3/abb2986_DataS4.csv) correspond with the paper. 

The mapping to the names used in this script is as follows:

- Cardiac mesoderm = ventral mesoderm
- Cranial mesoderm = cranial mesoderm
- Dorsal mesoderm = dorsal mesoderm
- Presomitic mesoderm = paraxial mesoderm

---

```r
## count matrix
data <- readRDS(paste0(dir, "data/mesodermData_reference.RAW.Rds"))

## sample metadata
meta <- read.csv(paste0(dir, "data/DataS4.csv"), header = TRUE, stringsAsFactors = FALSE)
meta <- meta[match(colnames(data), meta$cellID),]
stopifnot(identical(colnames(data), meta$cell))
meta$batch <- as.factor(paste0("batch_", meta$batch)) ## make 'batch' categorical

## gene information (from Ensembl version 87)
ann <- read.table(paste0(dir, "data/Mus_musculus.GRCm38.87.tsv"), sep="\t", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
colnames(ann) <- c("gene", "chr", "start", "end", "strand")
```

There are 748 cells in this dataset. These were processed on the same batch as the stage -1 cells from the `unbiased` dataset. 

Cells were collected from embryos from stages 1 and 2, and all regions but the PSM were collected for both time points.


```r
table(meta$regionAnn, stage=meta$stage)
```

```
##                  stage
##                     1   2
##   cardiacMesoderm  52 105
##   caudalPSM       110   0
##   cranialMesoderm 111 156
##   dorsalMesoderm   97 117
```

```r
## cardiac mesoderm from stage 1 is 'contaminated'
```


#### Quality control

First, we check several metrics for each library to determine bad quality samples and discard them.


```r
## remove the mapping stats from count data
mapping.stats <- data[49764:49768,]
data <- data[1:49763,]

## compute general QC metrics
qc <- data.frame(libSize=colSums(data), 
                 nGenes=apply(data, 2, function(x) length(which(x>0))),
                 mit=colSums(data[row.names(ann[ann$chr=="MT",]),]),
                 ercc=colSums(data[grep("ERCC",row.names(data)),]))

## plot
plots <- list()
## total counts in genes + spike-ins
plots[[1]] <- ggplot(qc, aes(x=as.factor(meta$regionAnn), y=log10(libSize+1))) + geom_violin() + geom_boxplot(width=0.05) + theme_classic() + theme(legend.position="none") + ylab(expression('log'[10]*' library size')) + xlab("batch") + geom_hline(yintercept = log10(50000), lwd=0.25, lty=2, col="grey") + ggtitle("total reads in genes") + theme(plot.title = element_text(face="bold", hjust=0.5))

## genes detected
plots[[2]] <- ggplot(qc, aes(x=as.factor(meta$regionAnn), y=nGenes)) + geom_violin() + geom_boxplot(width=0.05) + theme_classic() + theme(legend.position="none") + ylab("total genes") + xlab("batch") + geom_hline(yintercept = 4000, lwd=0.25, lty=2, col="grey") + ggtitle("number of genes detected") + theme(plot.title = element_text(face="bold", hjust=0.5))

## mitochondrial %
plots[[3]] <- ggplot(qc, aes(x=as.factor(meta$regionAnn), y=mit/libSize*100)) + geom_violin() + geom_boxplot(width=0.05) + theme_classic() + theme(legend.position="none") + ylab("% reads in MT genes") + xlab("batch") + geom_hline(yintercept = 10, lwd=0.25, lty=2, col="grey") + ggtitle("% reads in mitochondrial genes") + theme(plot.title = element_text(face="bold", hjust=0.5))

## spike-ins %
plots[[4]] <- ggplot(qc, aes(x=as.factor(meta$regionAnn), y=ercc/libSize*100)) + geom_violin(scale="width") + geom_boxplot(width=0.05) + theme_classic() + theme(legend.position="none") + ylab("% reads in spike-ins") + xlab("batch") + geom_hline(yintercept = 50, lwd=0.25, lty=2, col="grey") + ggtitle("% reads in ERCC spike-ins") + theme(plot.title = element_text(face="bold", hjust=0.5))

ggarrange(plotlist = plots, ncol = 2, nrow = 2)
```

![](07.1_referenceMesodermDataset_files/figure-html/qc-1.png)<!-- -->

Although most samples were sequenced successfully (libraries of more than 10^5 fragments), many contain almost exclusively reads from ERCC spike-in transcripts and very few endogenous genes. The distribution of reads in ERCC spike-ins is bimodal, with most cells having either ~8% or close to 100%; few cells show intermediate values. However, some of those in the 25-50% interval have large library sizes and thus still a large number of endogenous genes. Thus, we decide on liberal filters for the proportion of reads in ERCC spike-ins, with the largest determinant of good quality being the number of endogenous genes detected.

To pass quality-control, cells need to:

- Have more than 50,000 reads mapped to annotated genes.
- Have more than 4,000 genes detected.
- Have less than 10% of their reads assigned to mitochondrial genes.
- Have less than 50% of reads mapped to ERCC spike-ins.


```r
badQual <- which(qc$libSize < 5e4 | qc$mit/(qc$libSize+1)*100 >= 10 | qc$ercc/(qc$libSize+1)*100 >= 50 | qc$nGenes <= 4000)
# length(badQual) # 422 (56.42%)
```

With this criteria, 422 (56.42%) cells fail and are removed from downstream analyses.


```r
## remove bad quality samples from count matris
stopifnot(identical(row.names(qc), colnames(data)))
data <- data[,-badQual]
data <- data[rowSums(data)>0,] ## remove non-expressed genes

## and from metadata
stopifnot(identical(row.names(qc), meta$cellID))
meta <- meta[-badQual,]

stopifnot(identical(colnames(data), meta$cellID))

## save QC stats for future reference
qc$pass <- ifelse(row.names(qc) %in% colnames(data), 1, 0)
write.table(qc, file=paste0(dir, "data/QCstats_referenceMesodermCells.tsv"), quote = FALSE, sep="\t")
```

The clean dataset now consist of 326 cells that collectively express 27045 genes.

Whereas most regions retain around half of the total cells, only 25% of cranial mesoderm cells pass QC. However, there are still 68 cells to work with, since there were many more cells collected from this region.


```r
table(meta$regionAnn, stage=meta$stage)
```

```
##                  stage
##                    1  2
##   cardiacMesoderm 16 85
##   caudalPSM       58  0
##   cranialMesoderm 12 56
##   dorsalMesoderm  23 76
```

#### Normalisation

To normalise the data we use the method implemented in `scran`. Spike-ins are treated separately and normalised by total counts in spike-ins.


```r
## set up SingleCellExperiment object
genes <- grep("ENSMUSG", row.names(data), value = TRUE)
spikes <- grep("ERCC", row.names(data), value = TRUE)

# metadata row.names should be counts col.names
m <- meta
row.names(m) <- m$cellID
m$cellID <- NULL
stopifnot(identical(row.names(m), colnames(data)))

# add also gene info
ann <- ann[genes,]
stopifnot(identical(row.names(ann), row.names(data[genes,])))
# need info for spikes
tmp <- data.frame(gene=spikes, chr=paste0("ERCC",1:length(spikes)), start=1, end=2, strand=1, row.names = spikes)
ann <- rbind(ann, tmp)
stopifnot(identical(row.names(ann), row.names(data)))

## sce object
sce <- SingleCellExperiment(assays = list(counts=as.matrix(data)), colData = m, rowData = ann[,1:2])
## specify spike ins
is.spike <- grepl("^ERCC-", rownames(sce))
sce <- splitAltExps(sce, ifelse(is.spike, "ERCC", "gene"))

## normalisation
# plot(density(log10(rowMeans(counts(sce)))), bty="l", main="")
# abline(v=log10(1), lty=2)
## the default filter of mean>=1 is appropriate

## pre-cluster the data to protect the size factor estimation from too many DEGs
set.seed(0)
clusters  <- quickCluster(sce, min.size = 25, method = "igraph")

## estimate size factors
sce  <- computeSumFactors(sce, cluster = clusters, min.mean = 1)
sf <- sizeFactors(sce)
names(sf) <- colnames(counts(sce))
write.table(sf, file=paste0(dir, "data/sizeFactors_referenceDataset_minMean1.tsv"), quote = FALSE, sep = "\t")

plot(sf, colSums(counts(sce))/1e6, pch=16, xlab="size factors", ylab="library size (millions)", bty="l")
abline(lm((colSums(counts(sce))/1e6)~sf))
```

![](07.1_referenceMesodermDataset_files/figure-html/sizeFactors-1.png)<!-- -->

These size factors are used to normalise endogenous gene counts.


```r
sce <- logNormCounts(sce)
```


### Exploratory analysis

As a preliminary analysis, we examine the substructure in the data by visualising in low dimensional space, with UMAP.

The different anatomical regions generally segregate, except for the dorsal mesoderm and the caudal PSM which are intermingled.


```r
hvgs <- modelGeneCV2(sce)
# plot(hvgs$mean, hvgs$total, log="xy", col=ifelse(hvgs$FDR<0.05, 'red', 'black'))
# points(hvgs$mean, hvgs$trend, col="blue", pch=16, cex=0.5)

hvgs <- getTopHVGs(hvgs, var.field="ratio", n=2000)

## we remove mitochondrial and sexually dimorphic genes from the list of HVGs
remove <- rowData(sce)[rowData(sce)$chr %in% c("Y", "MT"),]
remove <- rbind(remove, rowData(sce)[rowData(sce)$gene=='Xist',])

hvgs <- setdiff(hvgs, row.names(remove))
write.table(hvgs, file=paste0(dir, "results/HVGs_referenceCells_minMean1_FDR0.05.tsv"), quote = FALSE, col.names = FALSE, row.names = FALSE)

set.seed(4957)
sce <- runUMAP(sce, subset_row = hvgs)

umap <- as.data.frame(reducedDim(sce, "UMAP"))
colnames(umap) <- c("x", "y")

## colour by batch
umap$plate <- as.factor(sce$plate)
umap$regionAnn <- as.factor(sce$regionAnn)
plot(umap$x, umap$y, pch=16, col=umap$regionAnn, xlab="", ylab="", axes=FALSE, main="anatomical region")
box(bty="l"); mtext(side=1, line=1, text = "UMAP - dim1"); mtext(side=2, line=1, text = "UMAP - dim2")
legend("topright", legend = levels(umap$regionAnn), col=1:4, pch=16, cex=0.75)
```

![](07.1_referenceMesodermDataset_files/figure-html/umap-1.png)<!-- -->

We also observe a small cluster of cells that seem to have quite different transcriptional profiles. This cluster contains primarily cardiac mesoderm cells. It is likely that, when dissecting out the cardiac mesoderm region, there was some contamination from the underlying endoderm, which would explain this very distinct subpopulation.



```r
saveRDS(sce, file=paste0(dir, "data/sce_referenceCells_goodQual.NORM.Rds"))

## normalised expression estimates
dataNorm <- logcounts(sce)
saveRDS(dataNorm, file = paste0(dir, "data/mesodermData_reference.goodQual.NORM.Rds"))
```



```r
sessionInfo()
```

```
## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggpubr_0.2.5                magrittr_1.5               
##  [3] RColorBrewer_1.1-2          scater_1.14.4              
##  [5] ggplot2_3.3.0               scran_1.14.5               
##  [7] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.0
##  [9] DelayedArray_0.12.0         BiocParallel_1.20.0        
## [11] matrixStats_0.55.0          Biobase_2.46.0             
## [13] GenomicRanges_1.38.0        GenomeInfoDb_1.22.0        
## [15] IRanges_2.20.1              S4Vectors_0.24.1           
## [17] BiocGenerics_0.32.0        
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.5.1            edgeR_3.28.0             BiocSingular_1.2.0      
##  [4] viridisLite_0.3.0        DelayedMatrixStats_1.8.0 RcppParallel_4.4.4      
##  [7] assertthat_0.2.1         statmod_1.4.32           dqrng_0.2.1             
## [10] GenomeInfoDbData_1.2.2   vipor_0.4.5              yaml_2.2.0              
## [13] pillar_1.4.2             lattice_0.20-38          glue_1.3.1              
## [16] limma_3.42.0             digest_0.6.23            XVector_0.26.0          
## [19] ggsignif_0.6.0           colorspace_1.4-1         cowplot_1.0.0           
## [22] htmltools_0.4.0          Matrix_1.2-18            pkgconfig_2.0.3         
## [25] zlibbioc_1.32.0          purrr_0.3.3              scales_1.1.0            
## [28] RSpectra_0.16-0          tibble_2.1.3             farver_2.0.1            
## [31] withr_2.1.2              crayon_1.3.4             evaluate_0.14           
## [34] FNN_1.1.3                beeswarm_0.2.3           tools_3.6.1             
## [37] lifecycle_0.1.0          stringr_1.4.0            munsell_0.5.0           
## [40] locfit_1.5-9.1           irlba_2.3.3              compiler_3.6.1          
## [43] rsvd_1.0.2               rlang_0.4.2              grid_3.6.1              
## [46] RCurl_1.95-4.12          BiocNeighbors_1.4.1      rstudioapi_0.10         
## [49] igraph_1.2.4.2           bitops_1.0-6             labeling_0.3            
## [52] rmarkdown_1.18           gtable_0.3.0             R6_2.4.1                
## [55] gridExtra_2.3            knitr_1.26               dplyr_0.8.3             
## [58] uwot_0.1.4               stringi_1.4.3            ggbeeswarm_0.6.0        
## [61] Rcpp_1.0.3               tidyselect_0.2.5         xfun_0.11
```
